xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -318,81 +501,45 @@ private LdsUpdate processServerSideListener(Listener listener)
             "HttpFilter [" + filterName + "] is not optional and has no typed config");
       }
     }
-    return parseRawFilterConfig(filterName, httpFilter.getTypedConfig(), isOptional, false);
-  }
-
-  @Nullable // Returns null if the filter should be ignored.
-  private static StructOrError<FilterConfig> parseRawFilterConfig(
-      String filterName, Any anyConfig, Boolean isOptional, boolean isOverrideConfig) {
-    checkArgument(
-        isOptional != null || isOverrideConfig, "isOptional can't be null for top-level config");
-    String typeUrl = anyConfig.getTypeUrl();
-    if (isOverrideConfig) {
-      isOptional = false;
-      if (typeUrl.equals(TYPE_URL_FILTER_CONFIG)) {
-        io.envoyproxy.envoy.config.route.v3.FilterConfig filterConfig;
-        try {
-          filterConfig =
-              anyConfig.unpack(io.envoyproxy.envoy.config.route.v3.FilterConfig.class);
-        } catch (InvalidProtocolBufferException e) {
-          return StructOrError.fromError(
-              "HttpFilter [" + filterName + "] contains invalid proto: " + e);
-        }
-        isOptional = filterConfig.getIsOptional();
-        anyConfig = filterConfig.getConfig();
-        typeUrl = anyConfig.getTypeUrl();
-      }
-    }
-    Message rawConfig = anyConfig;
-    if (anyConfig.getTypeUrl().equals(TYPE_URL_TYPED_STRUCT)) {
+    Message rawConfig = httpFilter.getTypedConfig();
+    String typeUrl = httpFilter.getTypedConfig().getTypeUrl();
+    if (typeUrl.equals(TYPE_URL_TYPED_STRUCT)) {
       TypedStruct typedStruct;
       try {
-        typedStruct = anyConfig.unpack(TypedStruct.class);
+        typedStruct = httpFilter.getTypedConfig().unpack(TypedStruct.class);
       } catch (InvalidProtocolBufferException e) {
         return StructOrError.fromError(
             "HttpFilter [" + filterName + "] contains invalid proto: " + e);
       }
       typeUrl = typedStruct.getTypeUrl();
       rawConfig = typedStruct.getValue();
     }
-    Filter filter = FilterRegistry.getDefaultRegistry().get(typeUrl);
-    if (filter == null) {
+    Filter filter = filterRegistry.get(typeUrl);
+    if (filter == null || (!isForClient && filter.isSupportedOnClients())
+        || (isForClient && filter.isSupportedOnServers())) {
```

Comment:
This expression should be:
```
    if (filter == null || (isForClient && !filter.isSupportedOnClients())
        || (!isForClient && !filter.isSupportedOnServers())) 
```

Basically: if I am processing for-client but the filter-is-not-supported-on-clients OR I am processing for-server but the filter-is-not-supported-on-servers

The 2 are not equivalent because `isSupportedOnClients()` and `isSupportedOnServers()` are not opposite of each other.

Reply:
Ahhh, good catch. `!isForClient` == isForServer, this is awkward. I wish I could use an enum instead of a boolean. But thought that's an overkill for such a small thing. I will fix and add some test.

Reply:
Fixed, and enabled Router filter for server side. It will be used for both client and server, and we can easily use it in the test to cover this logic. Also, I deleted the newly added interface methods `isSupportedOnClients()` and `isSupportedOnServers()`. Those are redundant, can just check with `instance of ClientInterceptorBuilder` and `instance of ServerInterceptorBuilder`. It's cumbersome to have two sources of truth.

Reply:
> Those are redundant, can just check with instance of ClientInterceptorBuilder and instance of ServerInterceptorBuilder. It's cumbersome to have two sources of truth.

Yes but `isSupported...` are more readable and I prefer those. These could have been made default methods in the interface with the correct Java version.